TAXII Collector bot and STIX Parser bot

laciKE opened the pull request:
As a bare minimum, TAXII Collector currently collects only the objects of type indicator. These objects contain information about indicators and the detection patterns, e.g. in stix, pcre, sigma, snort, suricata, yara format. The pattern, pattern_type and valid_from properties are required, while confidence, description and labels are only optional properties. However, they are present in several TAXII feeds and could be used to determine classification.taxonomy and classification.type even without processing the relationships of the indicators (e.g. indicator indicates malware)

STIX Parser is currently capable of parsing objects of type indicator (usually retrieved from the TAXII Collector). From the indicator objects, it extracts the detection pattern (currently only the single Observation Expressions in STIX format are supported). It supports IP addresses, Domains and URLs indicator values. Moreover, this parser also attempts to extract some optional properties of STIX objects such as description and labels, which can be useful for futher classification of the event with the Expert Bots

TAXII Collector tests for missing parameters and mock the simple TAXII server providing minimal collection with simple indicator object STIX Parser tests fo indicator patterns parsing
Improvements based on @sebix comments, collection title used as feed.code Fix codestyle in TAXII and STIX bots
Fix Python 3.8 support in STIX Parser bot.

laciKE committed:
Add TAXII Collector bot and STIX Parser bot

As a bare minimum, TAXII Collector currently collects only the objects of type indicator. These objects contain information about indicators and the detection patterns, e.g. in stix, pcre, sigma, snort, suricata, yara format. The pattern, pattern_type and valid_from properties are required, while confidence, description and labels are only optional properties. However, they are present in several TAXII feeds and could be used to determine classification.taxonomy and classification.type even without processing the relationships of the indicators (e.g. indicator indicates malware)

STIX Parser is currently capable of parsing objects of type indicator (usually retrieved from the TAXII Collector). From the indicator objects, it extracts the detection pattern (currently only the single Observation Expressions in STIX format are supported). It supports IP addresses, Domains and URLs indicator values. Moreover, this parser also attempts to extract some optional properties of STIX objects such as description and labels, which can be useful for futher classification of the event with the Expert Bots

TAXII Collector tests for missing parameters and mock the simple TAXII server providing minimal collection with simple indicator object
STIX Parser tests fo indicator patterns parsing
Improvements based on @sebix comments, collection title used as feed.code
Fix codestyle in TAXII and STIX bots
Fix Python 3.8 support in STIX Parser bot

laciKE commented:
The TAXII and STIX bots are currently tested with the ESET Threat Intelligence (ETI) feeds.
Recently, ETI added several new feeds which are available only via TAXII/STIX 2.1, and older ESETCollectorBot and ESETParserBot cannot handle them.

I am working on Expert Bot for classification events from ETI and I would like to publish it when it will be ready - together with feeds in `feeds.yaml`

laciKE commented:
Hello, I have a question regarding the proposal from the last commit.

I created ESETExpertBot which can add the proper classification.type and malware.name (if possible) to the events produced by StixParserBot. Ref: https://github.com/laciKE/intelmq/blob/eset/intelmq/bots/experts/eset/expert.py

When I wanted to add ESET Threat Intelligence TAXII feeds to feeds.yaml also with the expert bot, too, the tests failed, because it seems that the expert bot is not allowed in feeds.yaml.

Especially with the TAXII feeds, three bots will be needed to ingest those feeds:
- Collect STIX objects from TAXII server (generic TAXII Collector)
- Parse generic STIX indicator objects (generic STIX Parser)
- Apply vendor-specific enrichment of events based on optional STIX properties used by the particular vendor (vendor-specific Expert bot).

As far as I understand, two parsers cannot by chained in the pipeline (because the input is Report, and output is the Event). 
What is the suggested way to do a three-step ingestion in similar cases? One generic Parser bot for given format, and all vendor-specific bots should inherit from that generic parser bot?

sebix commented:
From what I understand, reading the code, the ESET expert fixes the classification for all events coming from the ESET feed. That logic should be in the Parser instead. Or is the code of ESET expert also useful for other sources other than ESET?

laciKE commented:
Thank you for your answer. You are right, that expert bot fixes the classification and it is ESET-specific. I will change it to parser bot, which will inherit from the StixParserBot from this pull request. After that, I will add the commits with "EsetStixParserBot" to this pull request.

sebix commented:
Ah, I see. That parser also works for multiple sources, other than ESET?

laciKE commented:
This StixParserBot yes, it should work for any source which provide Threat Intelligence data in STIX 2.1 format. I created it from scratch by reading [STIX 2.1 documentation](https://docs.oasis-open.org/cti/stix/v2.1/os/stix-v2.1-os.html), and it is able to parse [Indicators Objects](https://docs.oasis-open.org/cti/stix/v2.1/os/stix-v2.1-os.html#_muftrcpnf89v) with simple [Patterns](https://docs.oasis-open.org/cti/stix/v2.1/os/stix-v2.1-os.html#_me3pzm77qfnf).

StixParserBot (and TaxiiCollectorBot) should be used with any TAXII/STIX 2.1 feed. General parsing of indicators works, but for correct classification, the vendor-specific bot is needed. This is why I asked what is the proper way to do it. 

Currently I tested TaxiiCollectorBot+StixParserBot only with [ESET Threat Intelligence TAXII](https://help.eset.com/eti_portal/en-US/taxii_feeds.html) feeds, because I do not have access to other TAXII 2.1 feeds. For correct classification, I created the [ESETExpertBot](https://github.com/laciKE/intelmq/blob/eset/intelmq/bots/experts/eset/expert.py), which I am going to change to ESETStixParserBot (it will by child a class of generic StixParserBot)

laciKE committed:
Add ESET STIX Parser bot

Parser bot for enriching events from ESET Threat Intelligence, which were collected by TaxiiCollectorBot. It inherits from generic StixParserBot and implement vendor-specific parsing. ESET STIX Parser bot analyzes comment (based on original description of STIX Indicator object) and choose proper classification type and if possible, also fills the malware.name in the event.

laciKE committed:
Add ESET Threat Intelligence feeds

ETI feeds with URLs, domains and IP addresses, which can be collected by TaxiiCollectorBot and parsed by ESETStixParserBot

laciKE commented:
I will try to do better parsing for STIX2 patterns.

Also, in ESET Threat Intelligence there are sometimes domains reported in URL feed and IP addresses in Domain feed, and this causes InvalidValue exceptions in produced events - I will try to address it, at least by discarding those indicators without raising exceptions (raise_failure=False).

laciKE committed:
Improve STIX patterns parsing

Use the official STIX2 Pattern Validator to get thecomparison expressions and extracts simple IoCs from them. Support for URLs, Domains, IPv4, IPv6 and also for MD5, SHA-1 and SHA-256 hashes. Small fixes and workarounds implemented to address certain anomalies in STIX data provided by some vendors (e.g. ETI) - SHA1 and SHA256 keywords accepted, invalid objects reported as Domains or URLs are dropped without throwing the exceptions

laciKE committed:
Add TAXII and STIX bots documentation

laciKE commented:
Better parsing for STIX2 patterns ready, now the STIX parser bot can extract also hashes.

Above-mentioned issues with ESET Threat Intelligence fixed.

From my side, the PR is ready for review. If I should change something or if I forgot to do something, please, let me know, this is my first PR to IntelMQ.

laciKE committed:
Fix missing dependency error

sebix committed:
doc: add changelog entry for PR#2611

sebix commented:
I rebased on current develop and added the changelog entry. Review to come soon :)

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+
+    def parse_line(self, line, report):
+        """ Parse one STIX object of indicator type """
+        object_type = line.get('type', '')
```
Why is the fallback value of an empty string necessary?

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+                    yield event
+            else:
+                self.logger.warning('Unexpected type of pattern expression: %r, pattern: %r', pattern_type, pattern)
+        else:
```
Is it really OK to skip every other type?

if it is valid, it must at least be mentioned in the documentation.

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+                    self.parse_vendor_specific(event, line, report)
+                    yield event
+            else:
+                self.logger.warning('Unexpected type of pattern expression: %r, pattern: %r', pattern_type, pattern)
```
Same as with object types.

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+
+                    # IP address may be passed in Domain feeds or Domain may be passed in URL feeds
+                    # It violates the STIX format, however, in some sources it happens (e.g. in ETI)
+                    # Drop such events without failures and exceptions which slowing down the processing
```
Do you mean this?
```suggestion
                    # Drop such events without failures and exceptions, without slowing down the processing
```

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+                indicators = StixParserBot.parse_stix_pattern(pattern, self.logger)
+                for indicator_type, indicator_value in indicators:
+                    event = self.new_event(report)
+                    event.add('raw', json.dumps(line))
```
We can use the unparsed line by the json parser without the need of encoding it again:
```suggestion
                    event.add('raw', self.recover_line())
```

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+        """
+        if len(comparison) != 3:
+            if logger:
+                logger.warning('Unexpected Comparison Expressions. Expression: {}'.format(comparison))
```
```suggestion
                logger.warning('Unexpected Comparison Expressions. Expression: %r', comparison)
```

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+                                    ]
+        if not (property_name in supported_property_names) or (operator != '=') or not value.startswith("'") or not value.endswith("'"):
+            if logger:
+                logger.info('Unsupported Comparison Expression. Only Comparison Expressions with "equal" comparison operator and "value" or "hashes" property are supported. Expression: {}'.format(comparison))
```
```suggestion
                logger.info('Unsupported Comparison Expression. Only Comparison Expressions with "equal" comparison operator and "value" or "hashes" property are supported. Expression: %r', comparison)
```

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+            return
+
+        property_name, operator, value = comparison
+        supported_property_names = [['value'],
```
This is a constant and can be defined on module or class level.

Also: Can it be a tuples instead of lists?

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+                elif key == 'domain-name':
+                    indicators.append(('source.fqdn', value))
+                elif key == 'ipv4-addr':
+                    # remove port, sometimes the port is present in ETI
+                    value = value.split(':')[0]
```
Can we use the port for `source.port`?

sebix commented on intelmq/bots/parsers/stix/parser.py:
```diff
@@ -0,0 +1,149 @@
+                        indicators.append(('source.network', value))
+                    else:
+                        indicators.append(('source.ip', value))
+                elif key == 'file':
+                    if len(comparison) == 3 and len(comparison[0]) == 2 and comparison[0][0] == 'hashes':
+                        # converts MD5, SHA-1, SHA1, SHA-256, SHA256 to md5, sha1, sha256 used in IntelMQ
+                        hash_algo = comparison[0][1].lower().replace('-', '')
+                        indicators.append(('malware.hash.' + hash_algo, value))
```
Either ensure here that no other hashes than the ones supported by IntelMQ are in the STIX data (and not, e.g. SHA 512), _or_ in `parse_line` check if `indicator_type` is an allowed field, else add it to `extra.{indicator_type}`.

sebix commented on intelmq/tests/bots/collectors/taxii/test_collector_bot.py:
```diff
@@ -0,0 +1,108 @@
+# SPDX-FileCopyrightText: 2016 Sebastian Wagner, 2025 Ladislav Baco
```
```suggestion
# SPDX-FileCopyrightText: 2025 Ladislav Baco
```

sebix reviewed (changes requested):
Thank you for this thorough and well-done PR. High quality, also the tests.

I have a few remarks on some code lines.
